### PR TITLE
User login log out

### PR DESCRIPTION
**Pull request by nguyenthanhluanFramgia**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+        params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (hellovietnam93) [inline app/controllers/sessions_controller.rb]: using ""

Comment (hellovietnam93) [inline app/controllers/sessions_controller.rb]: ident

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,4 +1,9 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy,
+    :following, :followers]
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: where are they from?

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -21,8 +27,45 @@ def create
...
+      render :edit
+    end
+  end
+  
   private
   def user_params
     params.require(:user).permit :name, :email, :password, :password_confirmation
   end
+
+  def logged_in_user
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: puts in application controller to reuse

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -21,8 +27,45 @@ def create
...
+    unless logged_in?
+    store_location
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: ident

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -21,8 +27,45 @@ def create
...
+    flash[:danger] = t "please_log_in"
+    redirect_to login_url
+    end
+  end
+
+  def correct_user
+      @user = User.find params[:id]
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: ident

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -21,8 +27,45 @@ def create
     end
   end
 
+  def edit
+    @user = User.find params[:id]
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: why it need to use?

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -21,8 +27,45 @@ def create
...
+  end
+
+  def correct_user
+      @user = User.find params[:id]
+  end
+  
+  def admin_user
+      redirect_to root_url unless current_user.is_admin?
+  end
+
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: remove this line

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  def log_out
+    forget current_user
+    session.delete :user_id
+    @current_user = nil
+  end
+
+  def redirect_back_or default
+    redirect_to(session[:forwarding_url] || default)
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 9]
```diff
@@ -9,4 +9,46 @@ class User < ActiveRecord::Base
...
+  
+  before_save   :downcase_email
+
+  validates :name, presence: true, length: {maximum: 50}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 },
+    format: { with: VALID_EMAIL_REGEX },
+    uniqueness: { case_sensitive: false }
+  
+  def User.digest string
```
Comment (hellovietnam93) [inline app/models/user.rb]: search about class method convention

File: app/models/user.rb [hunk 10]
```diff
@@ -9,4 +9,46 @@ class User < ActiveRecord::Base
...
+
+  def forget
+    update_attribute :remember_digest, nil
+  end
+  
+   private
+   def downcase_email
+     self.email = email.downcase
+   end
+
```
Comment (hellovietnam93) [inline app/models/user.rb]: remove line

File: app/views/sessions/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,27 @@
+<% provide title: t("_login") %>
+<h1><%= t "_login" %> </h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
+
+      <%= f.label :email %>
+      <%= f.email_field :email, class: 'form-control' %>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: ""

File: app/views/sessions/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+      <%= f.label :password %>
+      <%= link_to t("forgot_password"), new_password_reset_path %>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: do it have in requirement?

File: config/routes.rb [hunk 13]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
```
Comment (hellovietnam93) [inline config/routes.rb]: ""

File: app/controllers/application_controller.rb [hunk 14]
```diff
@@ -1,5 +1,13 @@
 class ApplicationController < ActionController::Base
-  # Prevent CSRF attacks by raising an exception.
-  # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+  include SessionsHelper
+
+  private
+  def logged_in_user
+    unless logged_in?
+    store_location
```
Comment (hellovietnam93) [inline app/controllers/application_controller.rb]: ident

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -1,4 +1,8 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
+  before_action :correct_user, only: [:edit, :update]
+  before_action :admin_user, only: :destroy
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: don't need to check admin here, admin don't work in normal user side

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -14,6 +18,7 @@ def new
   def create
     @user = User.new user_params
     if @user.save
+      session[:user_id] = @user.id
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: call login function

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -21,8 +26,32 @@ def create
     end
   end
 
+  def destroy
+    User.find(params[:id]).destroy
+    flash[:success] = t "user_deleted"
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: if it cannot be destroy also notice deleted/

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -21,8 +26,32 @@ def create
...
+    redirect_to users_url
+  end
+
+  def update
+    @user = User.find params[:id]
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: why it need to find here

File: app/helpers/sessions_helper.rb [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+      user = User.find_by(id: user_id)
+      if user && user.authenticated?(:remember, cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
+
+  def logged_in?
+    !current_user.nil?
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: !current_user.nil? = current_user.present?

File: app/models/user.rb [hunk 20]
```diff
@@ -9,4 +9,45 @@ class User < ActiveRecord::Base
...
+  
+  before_save   :downcase_email
+
+  validates :name, presence: true, length: {maximum: 50}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 },
+    format: { with: VALID_EMAIL_REGEX },
+    uniqueness: { case_sensitive: false }
+  
+  def User.create_digest string
```
Comment (hellovietnam93) [inline app/models/user.rb]: what is class method convention?

File: app/views/sessions/new.html.erb [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    
+      <%= f.label :remember_me, class: "checkbox inline" do %>
+      <%= f.check_box :remember_me %>
+        <span> <%= t "remember_me" %></span>
+      <% end %>
+
+      <%= f.submit t("_login"), class: "btn btn-primary" %>
+    <% end %>
+
+    <p><%= t "new_user?"  %>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: one space baefore %>

File: config/routes.rb [hunk 22]
```diff
@@ -6,5 +6,6 @@
   get "login" => "sessions#new"
   post "login" => "sessions#create"
   delete "logout" => "sessions#destroy"
-  resources :users
+  resources :users 
+  resources :password_resets, only: [:new, :create, :edit, :update]
```
Comment (hellovietnam93) [inline config/routes.rb]: except

File: .byebug_history [hunk 23]
```diff
@@ -0,0 +1,11 @@
+continue
```
Comment (hellovietnam93) [inline .byebug_history]: remove this file, put it in .gitignore

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -21,8 +25,34 @@ def create
     end
   end
 
+  def edit
+  end
+
+  def destroy  
+    if @user.destroy
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: where is @user ?

File: app/helpers/sessions_helper.rb [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def current_user? user
+    user == current_user
+  end
+
+  def current_user
+    if (user_id = session[:user_id])
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+      user = User.find_by(id: user_id)
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+
+  def redirect_back_or default
+    redirect_to session[:forwarding_url] || default
+    session.delete :forwarding_url
+  end
+
+  def store_location
+    session[:forwarding_url] = request.url if request.get?
+  end
+
```
Comment (hellovietnam93) [inline app/helpers/sessions_helper.rb]: remove this line

File: app/models/user.rb [hunk 29]
```diff
@@ -9,4 +9,45 @@ class User < ActiveRecord::Base
   has_many :activities
 
   has_secure_password
+  validates :password, presence: true, length: {minimum: 6}, allow_nil: true
```
Comment (hellovietnam93) [inline app/models/user.rb]: put it in right block

File: app/models/user.rb [hunk 30]
```diff
@@ -9,4 +9,45 @@ class User < ActiveRecord::Base
...
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    remember_token = User.new_token
+    update_attribute(:remember_digest, User.create_digest(remember_token))
+  end
+
+  def authenticated? attribute, token
+    digest = send("#{attribute}_digest")
```
Comment (hellovietnam93) [inline app/models/user.rb]: remove ()

File: app/views/layouts/_header.html.erb [hunk 31]
```diff
@@ -6,7 +6,24 @@
         <li><%= link_to t("_home"), root_path %></li>
         <li><%= link_to t("_help"), help_path %></li>
         <li><%= link_to t("_about"), about_path %></li>
-        <li><%= link_to t("_users"), users_path %></li>        
+        <li><%= link_to t("_users"), users_path %></li> 
+        <% if logged_in? %>          
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              <%= t("account") %> <b class="caret"></b>
```
Comment (hellovietnam93) [inline app/views/layouts/_header.html.erb]: remove ()

File: app/views/sessions/new.html.erb [hunk 32]
```diff
@@ -0,0 +1,26 @@
+<% provide title: t("_login") %>
+<h1><%= t "_login" %> </h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: remove ()

File: app/controllers/users_controller.rb (hunk 24 above)
Comment (nguyenthanhluanFramgia) [inline app/controllers/users_controller.rb]: before_action :load_user

File: app/views/sessions/new.html.erb [hunk 33]
```diff
@@ -0,0 +1,26 @@
...
+
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+
+      <%= f.label :password %>      
+      <%= f.password_field :password, class: "form-control" %>
+    
+      <%= f.label :remember_me, class: "checkbox inline" do %>
+      <%= f.check_box :remember_me %>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: ident

File: app/views/users/edit.html.erb [hunk 34]
```diff
@@ -0,0 +1,9 @@
+<% provide title: t("_profile") %>
+<% provide :button_text, t("buttons.update_profile") %>
+<h1> <%= t "_signup" %></h1>
```
Comment (hellovietnam93) [inline app/views/users/edit.html.erb]: ```<h1><%= t "_signup" %></h1>```or ```<h1> <%= t "_signup" %> </h1>```
pay attention about convention.

File: .gitignore [hunk 35]
```diff
@@ -27,3 +27,4 @@ config/environments/*.local.yml
 
 # Ignore application configuration
 /config/application.yml
+.byebug_history
```
Comment (hellovietnam93) [inline .gitignore]: end file

File: app/controllers/users_controller.rb (hunk 24 above)
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: really?, plz check again?

File: app/controllers/users_controller.rb [hunk 36]
```diff
@@ -21,8 +25,34 @@ def create
     end
   end
 
+  def edit
+  end
+
+  def destroy  
+    if @user.destroy      
+      flash[:success] = t "user_deleted"
+      redirect_to users_url
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: write out of if else

File: app/models/user.rb [hunk 37]
```diff
@@ -9,4 +9,44 @@ class User < ActiveRecord::Base
...
 
   has_secure_password
+  attr_accessor :remember_token
+  
+  before_save :downcase_email
+  
+  validates :password, presence: true, length: {minimum: 6}, allow_nil: true
+  validates :name, presence: true, length: {maximum: 50}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 },
```
Comment (hellovietnam93) [inline app/models/user.rb]: convention remove space in {}

File: app/models/user.rb [hunk 38]
```diff
@@ -9,4 +9,44 @@ class User < ActiveRecord::Base
...
+    format: { with: VALID_EMAIL_REGEX },
+    uniqueness: { case_sensitive: false }
+  
+  def create_digest string
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+      BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
```
Comment (hellovietnam93) [inline app/models/user.rb]: remove ()

File: app/controllers/users_controller.rb (hunk 24 above)
Comment (nguyenthanhluanFramgia) [inline app/controllers/users_controller.rb]: tks

File: app/models/user.rb [hunk 39]
```diff
@@ -9,4 +9,44 @@ class User < ActiveRecord::Base
...
+  end
+
+  def new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    remember_token = User.new_token
+    update_attribute(:remember_digest, User.create_digest(remember_token))
```
Comment (hellovietnam93) [inline app/models/user.rb]: remove (), tell me the differences between update_attribute and update_attributes

File: app/models/user.rb [hunk 40]
```diff
@@ -9,4 +9,44 @@ class User < ActiveRecord::Base
...
+  end
+
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password? token
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
```
Comment (hellovietnam93) [inline app/models/user.rb]: using update_attributes

File: app/views/sessions/new.html.erb [hunk 41]
```diff
@@ -0,0 +1,21 @@
+<% provide title: t("_login") %>
+<h1><%= t "_login" %> </h1>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: 1 là sử dụng ```<h1> <%= t "_login" %> </h1>```
hoặc là ```<h1><%= t "_login" %></h1>```

File: app/views/sessions/new.html.erb [hunk 42]
```diff
@@ -0,0 +1,21 @@
...
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for :session, url: login_path do |f| %>
+    <%= f.label :email %>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: ident

File: app/models/user.rb (hunk 39 above)
Comment (nguyenthanhluanFramgia) [inline app/models/user.rb]: update_attribute: update 1 attribute
update_attributes: : update many attributes
i think that.

File: app/models/user.rb [hunk 43]
```diff
@@ -9,4 +9,44 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create string, cost: cost
+  end
+
+  def new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    remember_token = User.new_token
+    update_attributes(:remember_digest, User.create_digest remember_token)
```
Comment (hellovietnam93) [inline app/models/user.rb]: plz check it again. run again to see right or wrong

File: app/models/user.rb [hunk 44]
```diff
@@ -9,4 +9,44 @@ class User < ActiveRecord::Base
...
+  end
+
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password? token
+  end
+
+  def forget
+    update_attributes :remember_digest, nil
```
Comment (hellovietnam93) [inline app/models/user.rb]: like above

File: app/views/sessions/new.html.erb [hunk 45]
```diff
@@ -0,0 +1,21 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for :session, url: login_path do |f| %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>      
+      <%= f.password_field :password, class: "form-control" %>    
+      <%= f.label :remember_me, class: "checkbox inline" do %>
+      <%= f.check_box :remember_me %>
+      <span> <%= t "remember_me" %></span>
```
Comment (hellovietnam93) [inline app/views/sessions/new.html.erb]: ident

File: app/controllers/users_controller.rb [hunk 46]
```diff
@@ -1,4 +1,7 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
Comment (hellovietnam93) [inline app/controllers/users_controller.rb]: before_action :logged_in_user, except: [:show, :new, :create]

**Pull request merged**